components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/dao/impl/IdentityUserMetadataMgtDAOImpl.java

```diff
@@ -0,0 +1,508 @@
+    @Override
+    public void updateUserMetadata(UserStoreManager userStoreManager, String username,
+                                   String claimURI, String value, String eventName) throws IdentityEventException {
+
+        if (username.contains(IdentityRecoveryConstants.TENANT_ASSOCIATION_MANAGER)) {
+            return;
+        }
+
+        String tenantId;
+        try {
+            tenantId = String.valueOf(userStoreManager.getTenantId());
+        } catch (UserStoreException e) {
+            throw new IdentityEventException(
+                    String.format("Error occurred while retrieving the tenantId of the user related to %s event.",
+                            eventName), e);
+        }
+
+        String userStoreDomain = UserCoreUtil.extractDomainFromName(username);
+        if (StringUtils.isNotBlank(userStoreDomain) && IdentityRecoveryConstants.ASGARDEO_USER_DOMAIN_NAME
+                .equalsIgnoreCase(userStoreDomain) && isStoreIdentityClaimsInUserStoreEnabled(userStoreManager)) {
+            username = UserCoreUtil.removeDomainFromName(username);
+            storeMetadataToUserStore(userStoreManager, username, tenantId, claimURI, value, eventName);
+            return;
+        }
+
+        storeMetadataToIdentityStore(username, tenantId, claimURI, value, eventName);
+    }
+
+    @Override
+    public UserIdentityClaim loadUserMetadataFromCache(UserStoreManager userStoreManager, String userName) {
+
+        try {
+            if (userName != null) {
+                userName = UserCoreUtil.removeDomainFromName(userName);
+                if (!IdentityUtil.isUserStoreCaseSensitive(userStoreManager)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Case insensitive user store found. Changing username from : " + userName +
+                                " to : " + userName.toLowerCase(Locale.ENGLISH));
+                    }
+                    userName = userName.toLowerCase(Locale.ENGLISH);
+                } else if (!IdentityUtil.isUseCaseSensitiveUsernameForCacheKeys(userStoreManager)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Case insensitive username for cache key is used. Changing username from : "
+                                + userName + " to : " + userName.toLowerCase(Locale.ENGLISH));
+                    }
+                    userName = userName.toLowerCase(Locale.ENGLISH);
+                }
+
+                String domainName = userStoreManager.getRealmConfiguration().
+                        getUserStoreProperty(UserCoreConstants.RealmConfig.PROPERTY_DOMAIN_NAME);
+
+                IdentityDataStoreCacheKey key = new IdentityDataStoreCacheKey(domainName, userName);
+                int tenantId = userStoreManager.getTenantId();
+                UserIdentityClaim userIdentityDTO = identityDataStoreCache.getValueFromCache(key, tenantId);
+
+                if (userIdentityDTO != null && log.isDebugEnabled()) {
```

Comment:
Is the condition correct here? what happens when userIdentityDTO != null and debug hasn't been enabled?